typing/shape.ml

```diff
@@ -256,33 +265,123 @@ end) = struct
      actual substitutions, for example in [App(Abs(x, body), t)], when
      [v] is a thunk that will evaluate to the normal form of [t]. *)
 
+  (* Thunks contain lazy values, so they cannot safely be compared
+     using Stdlib.compare. We must define our own comparison functions
+     for normal forms and local environments, which can contain thunks. *)
+  let rec compare_nf (nf1 : nf) nf2 =
+    let c = Option.compare Uid.compare nf1.uid nf2.uid in
+    if c <> 0 then c else
+    compare_nf_desc nf1.desc nf2.desc
+  and compare_nf_desc d1 d2 =
+    match d1, d2 with
+    | NVar v1, NVar v2 ->
+        Ident.compare v1 v2
+
+    | NApp (nfa1, nfb1), NApp (nfa2, nfb2) ->
+        let c = compare_nf nfa1 nfa2 in
+        if c <> 0 then c else
+        compare_nf nfb1 nfb2
+
+    | NAbs (env1, x1, t1, thunk1), NAbs (env2, x2, t2, thunk2) ->
+        let c = Ident.compare x1 x2 in
+        if c <> 0 then c else
+        let c = compare_shape t1 t2 in
+        if c <> 0 then c else
+        let c = compare_local_env env1 env2 in
+        if c <> 0 then c else
+        compare_thunk thunk1 thunk2
+
+    | NStruct str1, NStruct str2 ->
+        Item.Map.compare compare_thunk str1 str2
+
+    | NProj (nf1, it1), NProj (nf2, it2) ->
+        let c = Item.compare it1 it2 in
+        if c <> 0 then c else
+        compare_nf nf1 nf2
+
+    | NLeaf, NLeaf ->
+        0
+
+    | NComp_unit s1, NComp_unit s2 ->
+        String.compare s1 s2
+
+    | NoFuelLeft d1, NoFuelLeft d2 ->
+        compare_shape_desc d1 d2
+
+    | (NVar _
+      | NApp _
+      | NAbs _
+      | NStruct _
+      | NProj _
+      | NLeaf
+      | NComp_unit _
+      | NoFuelLeft _
+      ), _ ->
+        let constructor = function
+          | NVar _ -> 0
+          | NApp _ -> 1
+          | NAbs _ -> 2
+          | NStruct _ -> 3
+          | NProj _ -> 4
+          | NLeaf -> 5
+          | NComp_unit _ -> 6
+          | NoFuelLeft _ -> 7
```

Comment:
Maybe this function could associate `0` to `NLeaf` and numbers `1-7` in the right order for the constructors with arguments? That way, the compiler could optimize the function using simple arithmetic on the block's tag instead of a more complex computation.

Reply:
Correct (this was my intent, but I forgot that there was a constant constructor in the mix), but I don't think it matters performance-wise. I'll do the change anyway, as it makes the code more natural.